### DISCUSSION TITLE
CharacterVirtual second SetShape() fails

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#158] [Q&A] [answered]
I have a character system that is working pretty well. 
I was initiating the crouch functionality but for some reason I can't get the shape to ever update.

I have two shapes: `standingShape` and `crouchingShape`

I have a function to set the shape:
```ts
set shape(shape: Jolt.Shape) {
  this.character.SetShape(
	  shape,
	  1.5 * this.physicsSystem.GetPhysicsSettings().mPenetrationSlop,
	  //@ts-ignore
	  this.filters.movingBPFilter,
	  this.filters.movingLayerFilter,
	  this.filters.bodyFilter,
	  this.filters.shapeFilter,
	  this.joltInterface.GetTempAllocator()
  );
}
```

When I initialize my class I call `this.shape = this.standingShape` and everything works as expected

However, if I call `this.shape = this.crouchingShape` any time after the initial setup, it does nothing.

I rewrote it to copy the original example:
```ts
const tryShape = this.character.SetShape(
		newShape,
		1.5 * this.physicsSystem.GetPhysicsSettings().mPenetrationSlop,
		//@ts-ignore
		this.filters.movingBPFilter,
		this.filters.movingLayerFilter,
		this.filters.bodyFilter,
		this.filters.shapeFilter,
		this.joltInterface.GetTempAllocator()
	);
	console.log("Shape set", tryShape);
```
It always fails..

I can initialize with the crouching shape or the standing shape, but I can never set it again once set. 
No error messages, no crashes. 

Any insights on what could cause this?

**Comment by DennisSmolek** [comment]
So I've gone through the code a bit. 

It seems if the shape isn't set, Jolt forces the set through, even if it would fail some collision test. 
If it does an update request, it does a collision test of the new shape (against what? and where?)


I'm copying the demo pretty close so I'm not seeing where I'm getting an offset, but if I manually shift the y-position 0.5 I clear the collision test:
```ts
const positionCrouching = new Jolt.Vec3(0, 0.5 * this.characterHeightCrouching + this.characterRadiusCrouching + 0.5, 0);
this.crouchingShape = new Raw.module.RotatedTranslatedShapeSettings(
			positionCrouching,
			rotation,
			new Jolt.CapsuleShapeSettings(
				0.5 * this.characterHeightCrouching,
				this.characterRadiusCrouching
			)).Create().Get();
```

I tried jumping, and even if the shape is airborn it fails. 
I'm going to do a test where there are no other bodies...
I do have a slight thing where the character starts slightly in the floor then pops out of it when it moves. I'm not sure it's related.
Again, I built this off the example so I'm not sure what could be off.

**Reply by LeXXik** [reply]
I am on the road, so can't check it, but I think it should fail, if your new shape penetrates the ground. This can happen if your character position (not the position of its collision shape) is a bit low and you don't accomodate it in your offset.

**Reply by DennisSmolek** [reply]
Another update:
So looking at this line: https://github.com/jrouwe/JoltPhysics/blob/3e13b2c097905c8ca77d98813389b56cc200932e/Jolt/Physics/Character/CharacterVirtual.cpp#L1155

I don't think the character can't swap shapes when inside a sensor that shares the same Broadphase/layer.
One cheater method I've used for bodies is to tag them with  `isSensor`. this seems to let them be ignored in collision testing. 
However, I don't think this tests for that. 

When I remove my other body and jump, I'm able to swap shapes. 

That means touching the floor blocks, and touching a sensor blocks.

**Reply by PhoenixIllusion** [reply]
Regarding the What and Where, even without the shape, the Character Virtual still has a Position and Velocity (just consider it to be a point-entity at that level). When you attempt to change the Shape, the system first does what you listed:
1. Is the existing Shape null, or physics system null? If so, we are not initialized and can change the shape to whatever we want.

2. Take the existing character Velocity and Position. Simulate a test of this new shape at the new position and velocity, and gather together anything that may be or is currently colliding with that space in the physics system.

This is done using the passed in filters and the character's internal pointer to the physics system itself.
If anything is intersecting the shape, check it against that Parameter 2 of SetShape (1.5 * mPenetrationSlop). If anything is intersecting deeper than that value, fail the shape-change and return false.

Regarding the offsets in general, and popping out of the ground: 
In the demo, the offsets are used to set the character's mPosition to the bottom of the capsule (like a characters feet). The mPosition is noted in the comments on CharacterVirtual.h to *not* be the center of a shape.
```cpp
// Current position (of the base, not the center of mass)
RVec3	mPosition = RVec3::sZero();
```

If you shapes you are assigning are not being offset by at least 1/2 their height and you are attempting to spawn the character on the ground, that may cause them to spawn in the ground and be pushed out.

If you have a small example of the test case, it would help in debugging any specific issues.

**Reply by PhoenixIllusion** [reply]
If this is only isolated to the use-case of attempting to change shape in the middle of a sensor, then you would be able to filter those out in the passed in SetShape filters (body/shape/layer) just for this situation, and then use the actual desired filters supporting sensor behavior in the rest of your simulated `ExtendedUpdate`. It looks like Pre-January the instructions were do not collide them in general, but Post-January this may be a bug to work around? https://github.com/jrouwe/JoltPhysics/commit/fb778c568d3ba14556559324671ffec172957f5c

**Comment by DennisSmolek** [comment]
> I am on the road, so can't check it, but I think it should fail, if your new shape penetrates the ground. This can happen if your character position (not the position of its collision shape) is a bit low and you don't accomodate it in your offset.

Which offset? The shape offset? I saw the property in the Jolt docs but the example doesn't set it anywhere...

**Comment by PhoenixIllusion** [comment, marked as answer]
Adding this at any section on the Example/character_virtual.html demo demonstrates the sensor issue:

```typescript
const sensorBox = createBox(new Jolt.RVec3(20, 4, 20), Jolt.Quat.prototype.sIdentity(), new Jolt.Vec3(4, 4, 4), Jolt.EMotionType_Static, LAYER_NON_MOVING, 0xFF00FF);
sensorBox.SetIsSensor(true);
const sensorBoxThree = dynamicObjects[dynamicObjects.length -1];
sensorBoxThree.material.wireframe = true;
```
![2024-04-24-character-virtual-sensorbox](https://github.com/jrouwe/JoltPhysics.js/assets/756488/9e043412-5a58-4d90-83a8-af4f6ed2137d)

I worked around it by using this just for SetShape on the bodyFilter parameter
```typescript
const bodyFilterIgnoreSensor = new Jolt.BodyFilterJS();
bodyFilterIgnoreSensor.ShouldCollide = () => true;
bodyFilterIgnoreSensor.ShouldCollideLocked = (inBody) => !Jolt.wrapPointer(inBody, Jolt.Body).IsSensor();
```
![2024-04-24-character-virtual-sensorbox-workaround](https://github.com/jrouwe/JoltPhysics.js/assets/756488/e46a1992-27eb-4a9d-8e8b-7361d6751680)

**Reply by DennisSmolek** [reply]
Adjusting the BodyFilter did the trick!
I must have another sensor body syncing with the character position and it was still blocking the collision. 

I use bodies as utility items to track in space and do things with constraints. I tried making them not collide using object filters but I didn't get the setup on that quite right. 
In the mean-time I used sensors as it meant they NEVER collided which is what I was really after. 

I've added your BodyFilter snipped for now until I pull the updated Jolt version. 

Thanks for all the fast work everyone!

**Comment by jrouwe** [comment]
Sorry about that, this is a bug. Setting a new shape should ignore sensors. I'll fix it immediately.

**Reply by jrouwe** [reply]
Should be fixed with 19b768f2983d89a30bb0ba6942983de697bf5b80
